jena-extras/jena-querybuilder/src/main/java/org/apache/jena/arq/querybuilder/rewriters/BuildElementVisitor.java

```diff
@@ -125,8 +125,16 @@ public void visit(ElementGroup el) {
             // noting to do
             result = el;
         } else if (lst.size() == 1) {
```

Comment:
@Claudenw What's the background to make the size=1 case special? (noting 2017 was a some time ago!)
It was introduced around Jena 3.7.0 which aligns with #2150.


I don't think the query builder should be making lst of 1 a special case.

There is no harm leaving groups of one. The parser generates them.

The algebra converter, query pretty printing (and OpAsQuery) all have to understand the group-of-one case. The algebra optimizer deals with them. The problem is that it does not deal with non-parser generated structures so well.

`ElementFilter` isn't the only example though the algebra converter copes with bare occurrences of other elements properly.

It is more valuable if the querybuilder generates ASTs like the parser. `Query.equals` then applies for the case `querybuilder -> string -> parse-> query` and `querybuilder.build -> query`.

If that then triggers querybuilder test failures, there may be a wider issue lurking.
(PS it does `WhereHandlerTest.testAddGraph` but `WhereValidator` seems to have its own idea of equality.

Reply:
It looks like a lot of the test suite assumes this simplification, removing it generates 67 failures of the module so possibly your approach from #2155 of allowing these forms to compile might be more broadly useful especially since users could be building query ASTs in other ways

@afs Just discovered while working on $dayjob code that's using query builder this morning that similar cases exists for other elements.  E.g. create a query with just a `BIND` also exhibits the same problem

Reply:
Then better to solve in the `AlgebraGenerator`.
I'll add `ElementBind` to #2155 and look for a general solution.

A separate matter which would need test case changes  is whether querybuilder should create ASTs the parser might generate and have the equivalance `querybuilder -> string -> parse-> query` and `querybuilder.build -> query`.

When used over the wire, queries from the querybuilder are turned into strings.

Issue #2156 for whether the querybuilder ought to try for parser-compatible syntax structures.

Reply:
I don't see a a general solution because syntax alone does not have a hierarchy that identifies elements into general categories (that's in the algebra).